### PR TITLE
Create webhooks-listener to listen to webhooks

### PR DESCRIPTION
**Pull request by apelisse**
Listens for webhooks, verify the signature, and pushes them (along with
event type) into a pub-sub queue.

All of this can be configured through a configmap: special path to push
along with github secret-key and topic to use.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1745)

### CODE REVIEW

**Comment by apelisse** [comment]
This is probably not complete, I could add tests and document exported functions. Let's have a first iteration of review though

**Comment by girishkalele** [comment]
What is the expected load/QPS on this webhooks handler ?

I see hardcoded 3 replicas - maybe an HPA would be better, scaling as needed ?

**Comment by apelisse** [comment]
Yeah, I'm very excited to do that. In a follow-up PR ;-)

**Comment by girishkalele** [comment]
Reviewed 4 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[mungegithub/hook-transform/hook.go, line 44 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1745#-KRjZn8axyfhcsW3NohZ:-KRjZn8axyfhcsW3Noh_:b-nezpal) ([raw file](https://github.com/kubernetes/contrib/blob/d8f55f6307a8776a38777bb382ee6d3f3516ca34/mungegithub/hook-transform/hook.go#L44)):*
> ```Go
> 	configFile string
> 	listenPort int
> 	projectId  string
> ```

Is this porjectId used ?

---

*[mungegithub/hook-transform/Makefile, line 8 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1745#-KRjXLhrZfPqIC1x4fgr:-KRjXLhrZfPqIC1x4fgs:b-ran3ma) ([raw file](https://github.com/kubernetes/contrib/blob/d8f55f6307a8776a38777bb382ee6d3f3516ca34/mungegithub/hook-transform/Makefile#L8)):*
> ```Makefile
> 
> build:
> 	CGO_ENABLED=0 go build .
> ```

This might not be reproducible for others - Go build needs Godeps frozen in a vendor/ directory or a Docker container that provides the build environment.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1745)*

**Comment by girishkalele** [comment]
Reviewed 1 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1745)*

**Comment by girishkalele** [comment]
Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[mungegithub/hook-transform/hook.go, line 119 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1745#-KRjZvpkbxdgsTCPn7H_:-KRjZvpkbxdgsTCPn7Ha:bfu2qaf) ([raw file](https://github.com/kubernetes/contrib/blob/d8f55f6307a8776a38777bb382ee6d3f3516ca34/mungegithub/hook-transform/hook.go#L119)):*
> ```Go
> 	w.Header().Set("Content-type", "text/plain")
> 	w.WriteHeader(http.StatusInternalServerError)
> 	w.Write([]byte(err.Error()))
> ```

Seems strange to send internal Gcloud error details back to Github.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1745)*

**Comment by girishkalele** [comment]
Godeps commit looks good...adding to existing mungegithub/vendor was good.

Code in hook.go LGTM.

File: mungegithub/webhooks-publisher/README.md [hunk 1]
```diff
@@ -14,3 +17,32 @@ kubectl create configmap webhook-config-v10 --from-file=config.yaml
 
 kubectl apply -f deployment.yaml
 ```
+
+How to use
+----------
+
+`config.yaml` contains the configuration:
+
+- `project` is the project that has the PubSub queue
```
Comment (girishkalele) [inline mungegithub/webhooks-publisher/README.md]: Is this the GCE Project Id ?
Is it true that your client must have GCE privileges to publish to that project ? Where are those credential ?

File: mungegithub/webhooks-publisher/README.md [hunk 2]
```diff
@@ -14,3 +17,32 @@ kubectl create configmap webhook-config-v10 --from-file=config.yaml
...
+- Each item in `paths` is the path it listens to webhooks and maps to the Github
+  `secret` and the PubSub `topic` where it should publish
+
+If you want to listen to a repository:
```
Comment (girishkalele) [inline mungegithub/webhooks-publisher/README.md]: How do we make Github send us the webhook notifications ? Is that documented somewhere too ?

**Review by girishkalele — COMMENTED** [review]
Some new comments on the docs

File: mungegithub/webhooks-publisher/README.md (hunk 2 above)
Comment (apelisse) [inline mungegithub/webhooks-publisher/README.md]: I've added a link to Github documentation

File: mungegithub/webhooks-publisher/README.md (hunk 1 above)
Comment (apelisse) [inline mungegithub/webhooks-publisher/README.md]: Yes, I've added a new section to explain.

**Comment by girishkalele** [comment]
New Documentation changes LGTM - cc @foxish @grodrigues3 for their comments.

File: mungegithub/webhooks-publisher/hook.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("Topic doesn't exist: %s", pathConfig.Topic)
+		}
+		handler := HookHandler{
+			Secret: pathConfig.Secret,
+			Topic:  topic,
+		}
+		glog.Infof("Setting up handler for %s (secret: %s): push to %s", path, pathConfig.Secret, pathConfig.Topic)
```
Comment (foxish) [inline mungegithub/webhooks-publisher/hook.go]: Should we log the secret?

**Comment by foxish** [comment]
LGTM

**Pull request merged**